bundles/model/org.openhab.model.item/src/org/openhab/model/item/binding/AbstractGenericBindingProvider.java

```diff
@@ -77,27 +76,31 @@ public void processBindingConfiguration(String context, Item item, String bindin
 			throw new BindingConfigParseException("context is not permitted to be null for item "+item);
 		}
 
-		Set<Item> items = contextMap.get(context);
-		if (items==null) {
-			items = new HashSet<Item>();
-			contextMap.put(context, items);
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if (items==null) {
+				items = new HashSet<Item>();
+				contextMap.put(context, items);
+			}
+
+			items.add(item);
 		}
-			
-		items.add(item);
 	}
 
 	/**
 	 * {@inheritDoc}
 	 */
 	public void removeConfigurations(String context) {
-		Set<Item> items = contextMap.get(context);
-		if(items!=null) {
-			for(Item item : items) {
-				// we remove all binding configurations for all items
-				bindingConfigs.remove(item.getName());
-				notifyListeners(item);
+		synchronized (contextMap) {
+			Set<Item> items = contextMap.get(context);
+			if(items!=null) {
+				for(Item item : items) {
+					// we remove all binding configurations for all items
+					bindingConfigs.remove(item.getName());
+					notifyListeners(item);
+				}
+				contextMap.remove(context);
 			}
```

Comment:
You have an internal/framework lock held when you make calls, via `notifyListeners`, to non-framework code.

It would be better if you bucketed _what_ to callout, while inside the framework lock, and then made the callouts themselves outside the lock...otherwise, depending upon the locking model of all of the called entities (Bindings), you could create a deadlock.


PS: You may want to run the Eclipse code-formatting function over the changes to align with the std openHAB uses.